nexus/src/app/sagas/instance_migrate.rs

```diff
@@ -276,35 +392,370 @@ async fn sim_v2p_ensure(
+                .body(None as Option<&serde_json::Value>)
+                .expect_status(Some(StatusCode::ACCEPTED)),
+        )
+        .authn_as(AuthnMode::PrivilegedUser)
+        .execute()
+        .await
+        .expect("Failed to stop instance");
+    }
+
+    fn select_first_alternate_sled(
+        db_instance: &db::model::Instance,
+        other_sleds: &[(Uuid, Server)],
+    ) -> Uuid {
+        let default_sled_uuid =
+            Uuid::parse_str(nexus_test_utils::SLED_AGENT_UUID).unwrap();
+        if other_sleds.is_empty() {
+            panic!("need at least one other sled");
+        }
+
+        if other_sleds.iter().any(|sled| sled.0 == default_sled_uuid) {
+            panic!("default test sled agent was in other_sleds");
+        }
+
+        if db_instance.runtime().sled_id == default_sled_uuid {
+            other_sleds[0].0
+        } else {
+            default_sled_uuid
+        }
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_saga_basic_usage_succeeds(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let other_sleds = add_sleds(cptestctx, 1).await;
+        let client = &cptestctx.external_client;
+        let nexus = &cptestctx.server.apictx().nexus;
+        let _project_id = setup_test_project(&client).await;
+
+        let opctx = instance_create::test::test_opctx(cptestctx);
+        let instance = create_instance(client).await;
+
+        // Poke the instance to get it into the Running state.
+        instance_simulate(cptestctx, nexus, &instance.identity.id).await;
+
+        let db_instance =
+            fetch_db_instance(cptestctx, &opctx, instance.identity.id).await;
+        let old_runtime = db_instance.runtime().clone();
+        let dst_sled_id =
+            select_first_alternate_sled(&db_instance, &other_sleds);
+        let params = Params {
+            serialized_authn: authn::saga::Serialized::for_opctx(&opctx),
+            instance: db_instance,
+            migrate_params: params::InstanceMigrate { dst_sled_id },
+        };
+
+        let dag = create_saga_dag::<SagaInstanceMigrate>(params).unwrap();
+        let saga = nexus.create_runnable_saga(dag).await.unwrap();
+        nexus.run_saga(saga).await.expect("Migration saga should succeed");
+
+        // Merely running the migration saga (without simulating any completion
+        // steps in the simulated agents) should not change where the instance
+        // is running.
+        let new_db_instance =
+            fetch_db_instance(cptestctx, &opctx, instance.identity.id).await;
+        assert_eq!(new_db_instance.runtime().sled_id, old_runtime.sled_id);
+        assert_eq!(
+            new_db_instance.runtime().propolis_id,
+            old_runtime.propolis_id
+        );
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_action_failure_can_unwind(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let log = &cptestctx.logctx.log;
```

Comment:
Thanks for adding this test!

Reply:
My pleasure! It's been an extremely useful backstop :)